Improve error boundary docs (#1100)
solidjs/solid-docs: src/routes/concepts/control-flow/error-boundary.mdx

order: 5
---

> By default, if part of your application throws an error during rendering, the entire application can crash, resulting in Solid removing its UI from the screen.

Reviewer: ```suggestion
By default, if part of an application throws an error during rendering, the entire application can crash, resulting in Solid removing its UI from the screen.
```

> Error boundaries provide a way to catch these errors and prevent the entire app from crashing.

Reviewer: ```suggestion
Error boundaries provide a way to catch the component(s) causing these errors and prevents the entire app from crashing.
```

Author: The error doesn’t always have to originate from a component. For example, as far as I know, when a `query` called with `createAsync` results in an error, `<ErrorBoundary>` captures that error:

```tsx
const user = createAsync(() => getUserAction());

<ErrorBoundary fallback={<div>Something went wrong!</div>}>
    {user()}
</ErrorBoundary>
```

Reviewer: Fair point.

The [`<ErrorBoundary>`](/reference/components/error-boundary) component is used to create an error boundary.

> `<ErrorBoundary>` catches any error that occurs during the rendering or updating of its children.

Reviewer: ```suggestion
It catches any error that occurs during the rendering or updating of its children.
```

> However, it's important to note that errors occurring outside the rendering process, such as in event handlers or after a `setTimeout`, are **not** caught by error boundaries.

Reviewer: ```suggestion
However, an important note is that errors occurring outside the rendering process, such as in event handlers or after a `setTimeout`, are *not* caught by error boundaries.
```

Reviewer: Could you also add back in the example - I think it's useful to have an example with fallback and without

Author: I get a Typescript error if I don't set a fallback:

```
Property 'fallback' is missing in type '{ children: string; }' but required in type '{ fallback: Element | ((err: any, reset: () => void) => Element); children: Element; }'. [2322]
```

It still works but it's obviously not the intended way to use `<ErrorBoundary>`.

> You can use the `fallback` prop to display a user-friendly error message or notification when an error occurs.

Reviewer: ```suggestion
The `fallback` prop can be used to display a user-friendly error message or notification when an error occurs.
```

> If you pass a function to `fallback`, it will receive the error object as well as a `reset` function.

Reviewer: ```suggestion
If a function id passed to `fallback`, it will receive the error object as well as a `reset` function.
```

> The `reset` function forces the `<ErrorBoundary>` to re-render its children and resets the error state, providing users with a way to recover from the error.

Reviewer: ```suggestion
The `reset` function forces the `<ErrorBoundary>` to re-render its children and reset the error state, providing users with a way to recover from the error.
```

```tsx
import { ErrorBoundary } from "solid-js";
import { Header, ErrorProne } from "./components";

function App() {
	return (
		<div>
			<Header />
			<ErrorBoundary
				fallback={(error, reset) => (
					<div>
						<p>Something went wrong: {error.message}</p>
						<button onClick={reset}>Try Again</button>
					</div>
				)}
			>
				<ErrorProne />
			</ErrorBoundary>
		</div>
	);
}
```

In this example, when the `ErrorProne` component throws an error, the `<ErrorBoundary>` catches it, preventing it from affecting the rest of the application.

> Instead, it displays an error message in place of the `ErrorProne` component.

Reviewer: ```suggestion
Instead, it displays the error message passed to the fallback prop.
```
